File: src/2024/iter-7(~2024-01-31)/198. House Robber/r2.js
var rob = function (nums) {
  if (nums.length === 0) {
    return 0;
  }
  if (nums.length === 1) {
    return nums[0];
  }
  if (nums.length === 2) {
    return Math.max(...nums);
  }

  let maxAtTwoBefore = nums[0];
  let maxAtOneBefore = Math.max(nums[0], nums[1]);

  for (let i = 2; i < nums.length; i++) {
    const maxAtCurrent = Math.max(nums[i] + maxAtTwoBefore, maxAtOneBefore);

    maxAtTwoBefore = maxAtOneBefore;
    maxAtOneBefore = maxAtCurrent;
  }

  return maxAtOneBefore;
};
// console.log(rob([2, 1, 1, 2]));
console.log(rob([2, 7, 9, 3, 1]));

File: src/2023/iter-4(~2023-10-31)/198. House Robber/r2.py
from typing import List


def rob(nums: List[int]) -> int:
    house_1 = 0
    house_2 = 0
    house_3 = 0
    for num in reversed(nums):
        temp = house_1
        house_1 = max(num + house_2, num + house_3)
        house_3 = house_2
        house_2 = temp

    return max(house_1, house_2)


print(rob([2, 7, 9, 3, 1]))
